### PR TITLE
Like param issue pull request

### LINKED ISSUE #771: LIKE parameter for zowe zos-files create data-set-*
**Is your feature request related to a problem or limitation? Please describe.**
LIKE parameter missing from zowe zos-files create data-set-*

**Describe the experience you'd like**
I would like to have --LIKE parameter for create data-set functions 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Nike682631**
Fix #771

### CODE REVIEW

File: packages/cli/src/zosfiles/-strings-/en.ts
Comment (t1m0thyj) [inline packages/cli/src/zosfiles/-strings-/en.ts:143]: Please remove trailing spaces

File: packages/cli/src/zosfiles/create/Create.options.ts
Comment (t1m0thyj) [inline packages/cli/src/zosfiles/create/Create.options.ts:58]: Please add one more space in front of these lines to make linter happy 😋

File: packages/cli/src/zosfiles/create/like/like.definition.ts
Comment (t1m0thyj) [inline packages/cli/src/zosfiles/create/like/like.definition.ts:56]: Original comment from @awharn:
> For the initial implementation of this command, I believe the only option should be `ZosFilesCreateExtraOptions.like`.

Please remove lines 42-55. The `CreateDefaults` import at top of file should also be unnecessary after making this change and can be removed.

**Review by t1m0thyj — CHANGES_REQUESTED** [review]
Thanks for addressing comments on the first PR 👍 I've left a few more - after addressing them hopefully the build will pass 🙂 

Please remember to add some unit, integration, and system tests (found in `packages/cli/__tests__/zosfiles`).

File: packages/cli/__tests__/zosfiles/__system__/create/ds/__scripts__/create_ds_help.sh
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__system__/create/ds/__scripts__/create_ds_help.sh:5]: Please remove the arguments "my.data.copy --like my.data.orig --lrecl 1024". Only the "--help" argument should be passed for this test.

File: packages/cli/__tests__/zosfiles/__unit__/create/like/__snapshots__/ds.definition.unit.test.ts.snap
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__unit__/create/like/__snapshots__/ds.definition.unit.test.ts.snap:1]: Please rename this folder from "create/like" to "create/ds" to make it consistent with the command name. Sorry for not noticing this earlier.

File: packages/cli/__tests__/zosfiles/__unit__/create/like/ds.definition.unit.test.ts
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__unit__/create/like/ds.definition.unit.test.ts:16]: Please update the directory in path for your new handler: "../../../../../src/zosfiles/create/ds/ds.definition" (after renaming "create/like" to "create/ds" as mentioned below).

File: packages/cli/__tests__/zosfiles/__unit__/create/like/ds.handler.unit.test.ts
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__unit__/create/like/ds.handler.unit.test.ts:19]: Please update the directory in path for your new handler: "../../../../../src/zosfiles/create/ds/ds.handler" (after renaming "create/like" to "create/ds" as mentioned below).

File: packages/cli/__tests__/zosfiles/__unit__/create/like/ds.handler.unit.test.ts
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__unit__/create/like/ds.handler.unit.test.ts:94]: DATA_SET_CLASSIC should be DATA_SET_LIKE here

File: packages/cli/src/zosfiles/create/like/ds.handler.ts
Comment (t1m0thyj) [inline packages/cli/src/zosfiles/create/like/ds.handler.ts:20]: Please rename to applicable name like "DataSetHandler"

**Review by t1m0thyj — CHANGES_REQUESTED** [review]
Thanks for adding tests! I've left a few comments that should help to get them passing 🙂

File: packages/cli/__tests__/zosfiles/__unit__/create/like/ds.handler.unit.test.ts
Comment (Nike682631) [reply packages/cli/__tests__/zosfiles/__unit__/create/like/ds.handler.unit.test.ts:94]: DATA_SET_LIKE does not exist in createdatasettypeenum

Comment (t1m0thyj) [reply packages/cli/__tests__/zosfiles/__unit__/create/like/ds.handler.unit.test.ts]: Sorry about that 😢 In this case you'll want to test that `Create.dataSetLike` was called rather than `Create.dataSet`. Please also update the expected parameters to match `Create.dataSetLike(session: AbstractSession, dataSetName: string, likeDataSetName: string)`.

File: packages/cli/__tests__/zosfiles/__unit__/create/like/ds.definition.unit.test.ts
Comment (t1m0thyj) [reply packages/cli/__tests__/zosfiles/__unit__/create/like/ds.definition.unit.test.ts]: Test here still failing because "classicPds" in the path needs to be replaced with "ds".

**Review by t1m0thyj — CHANGES_REQUESTED** [review]
Thanks for addressing comments 👍 

Integration tests appear to be failing because the new "create ds" command isn't recognized by Zowe CLI. Please register the new command in the "files create" group by adding `DsDefinition` to the list in this file: https://github.com/zowe/zowe-cli/blob/master/packages/cli/src/zosfiles/create/Create.definition.ts#L37-L46

File: packages/cli/__tests__/zosfiles/__unit__/create/ds/ds.handler.unit.test.ts
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__unit__/create/ds/ds.handler.unit.test.ts:95]: Please change `Create.dataSet` to `Create.dataSetLike` as done on the line below. `Create.dataSetLike` is the only SDK method being called by the "create ds" command and mocked for this test.

File: packages/cli/src/zosfiles/create/ds/ds.definition.ts
Comment (t1m0thyj) [inline packages/cli/src/zosfiles/create/ds/ds.definition.ts:28]: Please update from "like.handler" to "ds.handler" since the file was renamed.

File: packages/cli/src/zosfiles/create/ds/ds.handler.ts
Comment (t1m0thyj) [inline packages/cli/src/zosfiles/create/ds/ds.handler.ts:25]: Please change to `commandParameters.arguments.like`. There originally was a positional called "likeDataSetName", but it was changed into the "--like" option which can be accessed with `arguments.like`.

File: packages/cli/__tests__/zosfiles/__system__/create/ds/__scripts__/command/command_create_ds_like.sh
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__system__/create/ds/__scripts__/command/command_create_ds_like.sh:5]: Please parameterize this line so the system test can pass values into the script. Something like this should work:
`zowe files create data-set "$1.test.data.set.like" --like $2`

For a similar example, see `packages/cli/__tests__/zosfiles/__system__/create/binaryPds/__scripts__/command/command_create_binary_pds.sh`

File: packages/cli/__tests__/zosfiles/__system__/create/ds/__snapshots__/cli.files.create.dataset.classic.system.test.ts.snap
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__system__/create/ds/__snapshots__/cli.files.create.dataset.classic.system.test.ts.snap:1]: Please delete this file since there isn't a system test anymore that matches its name.

File: packages/cli/__tests__/zosfiles/__system__/create/ds/cli.files.create.ds.system.test.ts
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__system__/create/ds/cli.files.create.ds.system.test.ts:65]: Please define a suffix here (for example `dsnameSuffix = "like"`), so that the condition on line 58 will evaluate to true and the data set will be cleaned up after test finishes running.

File: packages/cli/__tests__/zosfiles/__system__/create/ds/cli.files.create.ds.system.test.ts
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__system__/create/ds/cli.files.create.ds.system.test.ts:67]: In addition to `user`, please supply a second parameter with the name of a dataset to "allocate like" from. For example, you could use `defaultSystem.zosjobs.iefbr14PSDataSet` which is a data set already defined to exist in the system test properties.

**Review by t1m0thyj — CHANGES_REQUESTED** [review]
Tests are getting close to being finished and passing 🙂 

Please address the comments, and then run tests with `-u` flag to update snapshots:
```shell
npm run test:unit -- -u
npx jest cli.files.create.dataset.integration.test.ts -u
npx jest cli.files.create.ds.system.test.ts -u
```

If you don't have access to a mainframe and are unable to run the last command to update system test snapshots, that's ok - just let me know and I can run the system test for you.

**Comment by Nike682631** [comment]
> Tests are getting close to being finished and passing 
> 
> Please address the comments, and then run tests with `-u` flag to update snapshots:
> 
> ```shell
> npm run test:unit -- -u
> npx jest cli.files.create.dataset.integration.test.ts -u
> npx jest cli.files.create.ds.system.test.ts -u
> ```
> 
> If you don't have access to a mainframe and are unable to run the last command to update system test snapshots, that's ok - just let me know and I can run the system test for you.

yes i don't have access of mainframe, please do that part for me.

File: packages/cli/__tests__/zosfiles/__integration__/create/ds/__scripts__/command/command_create_ds_fail_missing_like.sh
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__integration__/create/ds/__scripts__/command/command_create_ds_fail_missing_like.sh:5]: Please add host, user, and password arguments to this command, so that it skips prompting for those values and fails due to missing like data set name as desired. For example:
`zowe files create data-set my.data.copy --host fakehost --user fakeuser --pw fakepass`

File: packages/cli/__tests__/zosfiles/__unit__/create/ds/ds.handler.unit.test.ts
Comment (t1m0thyj) [inline packages/cli/__tests__/zosfiles/__unit__/create/ds/ds.handler.unit.test.ts:59]: This is shorthand for `likeDataSetName: likeDataSetName`. Since the command was reworked to have a "--like" argument instead of a "likeDataSetName" positional, it needs to be called "like" instead. Please change to `like: likeDataSetName`

**Comment by gejohnston** [comment]
There is no description of the expected behavior of this PR and no reference to the issue driving this PR.
After searching it looks like issue #771 is the relevant issue.

After reading code and the issue, I have one remaining question: What is the behavior when no --like option is supplied?

    zowe files create new.data.set

**Comment by Nike682631** [comment]
> There is no description of the expected behavior of this PR and no reference to the issue driving this PR.
> After searching it looks like issue #771 is the relevant issue.
> 
> After reading code and the issue, I have one remaining question: What is the behavior when no --like option is supplied?
> 
> ```
> zowe files create new.data.set
> ```

@gejohnston It should give out an error i believe

File: packages/cli/src/zosfiles/create/ds/ds.definition.ts
Comment (t1m0thyj) [inline packages/cli/src/zosfiles/create/ds/ds.definition.ts:49]: Please remove the default values for these options - the "create ds" command shouldn't have any as mentioned in #771:
> Now, for this zowe files create data-set command, I'm thinking all of the options should be left without defaults. The expectation being that users will either leverage the --like option and specify optional overrides or not leverage the --like option and fully specify the attributes. If they didn't want to fully specify the attributes they could always opt to use one of the template type creation commands.

File: packages/zosfiles/src/methods/create/Create.ts
Comment (t1m0thyj) [inline packages/zosfiles/src/methods/create/Create.ts:140]: I like this approach to add optional argument so there isn't a breaking change 👍

File: packages/zosfiles/src/methods/create/Create.ts
Comment (t1m0thyj) [inline packages/zosfiles/src/methods/create/Create.ts:143]: Please validate the options before calling the ZosmfRestClient. Example of doing this in `Create.dataSet`:
https://github.com/zowe/zowe-cli/blob/6b993c90093e84ab3af7a0e281fda7d8b86fdff7/packages/zosfiles/src/methods/create/Create.ts#L123-L125

Comment (t1m0thyj) [inline packages/zosfiles/src/methods/create/Create.ts:143]: Only the "like" option is being passed to the REST client: `JSON.stringify({ like: likeDataSetName })`
Other options in the `options` object must also be passed in. Please modify to stringify those options as well. Something like this should work: `{ like: likeDataSetName, ...(options || {}) }`

**Review by t1m0thyj — CHANGES_REQUESTED** [review]
Thanks for adding the requested CLI options 🙂 Left a few comments.

To fix outdated snapshots in tests, please run these commands:
Unit tests: `npm run test:unit -- -u`
Integration tests: `npx jest <testFilename> -u`

File: packages/zosfiles/src/methods/create/Create.ts
Comment (t1m0thyj) [inline packages/zosfiles/src/methods/create/Create.ts:150]: The `Create.datasetLike` API appears to be calling the z/OSMF endpoint twice. Please delete lines 153-155, and move lines 148-150 inside the `try` block. This should result in the desired behavior to call the endpoint only once with `tempOptions`.

File: packages/cli/src/zosfiles/create/ds/ds.definition.ts
Comment (t1m0thyj) [inline packages/cli/src/zosfiles/create/ds/ds.definition.ts:62]: The `options` property doesn't support an array with multiple values. Please add the second example as a new `{description: string, options: string}` object in the `examples` array.

**Review by t1m0thyj — CHANGES_REQUESTED** [review]
Thanks for making the requested updates 🙂 

Before I can try running the `create ds` command on the mainframe, there are a few more issues to address.

File: packages/cli/src/zosfiles/create/ds/ds.definition.ts
Comment (t1m0thyj) [reply packages/cli/src/zosfiles/create/ds/ds.definition.ts:62]: After making this change, please update unit and integration test snapshots to reflect the updated command help. 🙂

File: packages/cli/src/zosfiles/-strings-/en.ts
Comment (t1m0thyj) [inline packages/cli/src/zosfiles/-strings-/en.ts:47]: Please change `lrect` to `lrecl`, and update unit and integration test snapshots after making this change.

File: packages/zosfiles/src/methods/create/Create.ts
Comment (t1m0thyj) [inline packages/zosfiles/src/methods/create/Create.ts:147]: The `Create.dataSetValidateOptions` API doesn't recognize "like" as a valid option. Please add a case for "like" (e.g., `case "like":`) to the `switch` statement below:
https://github.com/zowe/zowe-cli/blob/7c16e6587628ce86f26ecc60bd64356df4ac4a02/packages/zosfiles/src/methods/create/Create.ts#L296-L301

**Review by t1m0thyj — APPROVED** [review]
Looks good to me, just needs 1 more approval and a vulnerability fix which we can handle before merging 🙂

**Comment by Nike682631** [comment]
> Looks good to me, just needs 1 more approval and a vulnerability fix which we can handle before merging slightly_smiling_face

finally :D

**Review by zFernand0 — APPROVED** [review]
LGTM! 😋

**Review by awharn — APPROVED** [review]

**Review by zFernand0 — APPROVED** [review]
Thanks for the contribution!
Good to go!

**Pull request merged**
